processor/processor.go

```diff
@@ -735,67 +944,99 @@ func (proc *HandleT) processJobsForDest(jobList []*jobsdb.JobT, parsedEventList
 				response = proc.transformer.Transform(eventList, integrations.GetUserTransformURL(proc.processSessions), userTransformBatchSize, false)
 				destStat.userTransform.End()
 			}
-			for _, userTransformedEvent := range response.Events {
-				// Fix this: Define a cleaner API. Custom transformer should not try to mimic internal structs
-				message := types.SingularEventT(userTransformedEvent.Output["message"].(map[string]interface{}))
-				updatedEvent := transformer.TransformerEventT{
-					Message:     message,
-					Metadata:    metadata,
-					Destination: destination,
-				}
-				eventsToTransform = append(eventsToTransform, updatedEvent)
-			}
+
+			eventsToTransform, failedJobs = proc.getDestTransformerEvents(response, metadata, destination, eventsByMessageID)
+			procErrorJobs = append(procErrorJobs, failedJobs...)
 			logger.Debug("Custom Transform output size", len(eventsToTransform))
 		} else {
 			logger.Debug("No custom transformation")
 			eventsToTransform = eventList
 		}
-		logger.Debug("Dest Transform input size", len(eventsToTransform))
-		destStat.destTransform.Start()
-		response = proc.transformer.Transform(eventsToTransform, url, transformBatchSize, false)
-		destStat.destTransform.End()
 
-		destTransformEventList := response.Events
-		logger.Debug("Dest Transform output size", len(destTransformEventList))
-		if !response.Success {
-			logger.Debug("[Processor: processJobsForDest] Request to transformer not a success ", response.Events)
-			continue
-		}
-		destStat.numOutputEvents.Count(len(destTransformEventList))
+		if len(eventsToTransform) > 0 {
+			logger.Debug("Dest Transform input size", len(eventsToTransform))
+			destStat.destTransform.Start()
+			response = proc.transformer.Transform(eventsToTransform, url, transformBatchSize, false)
+			destStat.destTransform.End()
 
-		//Save the JSON in DB. This is what the router uses
-		for _, destEvent := range destTransformEventList {
-			destEventJSON, err := json.Marshal(destEvent.Output)
-			//Should be a valid JSON since its our transformation
-			//but we handle anyway
-			if err != nil {
+			destTransformEventList := response.Events
+			logger.Debug("Dest Transform output size", len(destTransformEventList))
+			if !response.Success {
+				logger.Debug("[Processor: processJobsForDest] Request to transformer not a success ", response.Events)
 				continue
 			}
+			destStat.numOutputEvents.Count(len(destTransformEventList))
+
+			for _, failedEvent := range response.FailedEvents {
+				var payload json.RawMessage
+				if len(failedEvent.Metadata.MessageIDs) > 0 {
+					messageIds := failedEvent.Metadata.MessageIDs
+					var x []types.SingularEventT
+					for _, msgID := range messageIds {
+						x = append(x, eventsByMessageID[msgID])
+					}
+					var err error
+					payload, err = json.Marshal(x)
+					if err != nil {
+						panic(err)
```

Comment:
Instead of panic, let's log an error and move on? If the user transformation returns some invalid JSON, it is risky to crash the server here.

Reply:
Should not happen because the payload is from eventsByMessageID which are the ones stored in gw_jobs. So, should be valid json in postgres.
Removed the panic anyways